File: LiberatePDF2-Rest-Service/src/main/java/de/huwi/liberatepdf2/restservice/restrictionsremover/pdftk/legacy/LegacyPdftkRestrictionsRemover.java
package de.huwi.liberatepdf2.restservice.restrictionsremover.pdftk.legacy;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.file.Path;
import java.util.ArrayList;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import de.huwi.liberatepdf2.restservice.restrictionsremover.RestrictionsRemoverService;
import de.huwi.liberatepdf2.restservice.storage.FilesystemStorageService;

public class LegacyPdftkRestrictionsRemover {

	private final Logger Logger = LoggerFactory.getLogger(RestrictionsRemoverService.class);

	public File RemoveRestrictions(final File fileOriginal, final String password) {
		this.Logger.info("Filename = \"{}\"", fileOriginal);
		this.Logger.info("Password = \"{}\"", password);

		// String exePath = "C:\\Program Files (x86)\\PDFtk\\bin\\pdftk.exe";
		// String exePath = "/usr/bin/pdftk";
		final String pdftkExecutable = "pdftk";
		final File fileNew = new File(fileOriginal + FilesystemStorageService.SUFFIX_PDF_UNRESTRICTED);
		this.Logger.info("New filename = \"{}\"", fileNew);

		try {
			String stdoutLog = "";
			String stderrLog = "";
			String s;

			// build command
			ProcessBuilder processBuilder;
			if ((password != null) && !password.isEmpty()) {
				String passwordArgument = "";

				passwordArgument = "input_pw";
				processBuilder = new ProcessBuilder(pdftkExecutable, fileOriginal.getAbsolutePath(), passwordArgument,
						password, "output", fileNew.getAbsolutePath(), "allow", "AllFeatures");
			} else {
				processBuilder = new ProcessBuilder(pdftkExecutable, fileOriginal.getAbsolutePath(), "output",
						fileNew.getAbsolutePath(), "allow", "AllFeatures");
			}

			// run command
			final Process p = processBuilder.start();
			this.Logger.info("Command = \"{}\"", processBuilder.command());

			final BufferedReader stdOutput = new BufferedReader(new InputStreamReader(p.getInputStream()));
			final BufferedReader stdError = new BufferedReader(new InputStreamReader(p.getErrorStream()));

			// read the output from the command
			while ((s = stdOutput.readLine()) != null) {
				stdoutLog += s + "\n";
			}

			// read any errors from the attempted command
			while ((s = stdError.readLine()) != null) {
				stderrLog += s + "\n";
			}

			this.Logger.info("Command standard output: {}" + stdoutLog);
			this.Logger.info("Command error output: {}" + stderrLog);
		} catch (final IOException e) {
			this.Logger.error("Something went wrong during converting PDF", e);
			System.exit(-1);
		}

		return fileNew;
	}

	public Path[] RemoveRestrictions(final Iterable<Path> filesOriginal, final String password) {
		this.Logger.info("Removing restrictions from multiple files");
		final ArrayList<Path> filesNew = new ArrayList<>();

		for (final Path fileOriginal : filesOriginal) {
			final Path fileNew = this.RemoveRestrictions(fileOriginal.toFile(), password).toPath();
			filesNew.add(fileNew);
		}

		return filesNew.toArray(new Path[] {});
	}
}

File: microservice/src/main/kotlin/de/debuglevel/liberatepdf2/restservice/restrictionsremover/pdftk/legacy/LegacyPdftkRestrictionsRemover.kt
package de.debuglevel.liberatepdf2.restservice.restrictionsremover.pdftk.legacy

import mu.KotlinLogging
import org.apache.commons.io.IOUtils
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Path

class LegacyPdftkRestrictionsRemover {
    private val logger = KotlinLogging.logger {}

    val SUFFIX_PDF_UNRESTRICTED = ".unrestricted.pdf"

    private fun buildProcess(
        restrictedPdfPath: Path?, password: String?,
        unrestrictedPdfPath: Path
    ): ProcessBuilder {
        logger.debug { ("Building PDFtk process") }

        // String exePath = "C:\\Program Files (x86)\\PDFtk\\bin\\pdftk.exe";
        // String exePath = "/usr/bin/pdftk";
        val pdftkExecutable = "pdftk"
        val processBuilder: ProcessBuilder
        processBuilder = if (password != null && password.isNotEmpty()) {
            logger.debug { ("No password given or password is empty, building command without password argument") }
            val passwordArgument = "input_pw"
            ProcessBuilder(
                pdftkExecutable, restrictedPdfPath!!.toAbsolutePath().toString(),
                passwordArgument, password, "output", unrestrictedPdfPath.toAbsolutePath().toString(), "allow",
                "AllFeatures"
            )
        } else {
            logger.debug { ("Password given, building command with password argument") }
            ProcessBuilder(
                pdftkExecutable, restrictedPdfPath!!.toAbsolutePath().toString(),
                "output", unrestrictedPdfPath.toAbsolutePath().toString(), "allow", "AllFeatures"
            )
        }
        logger.debug { ("Final PDFtk command is: {}" + java.lang.String.join(" ", processBuilder.command())) }
        return processBuilder
    }

    @Throws(IOException::class)
    private fun readOutput(process: Process) {
        val outputInputStream = process.inputStream
        val errorInputStream = process.errorStream
        val stdoutLog = IOUtils.toString(outputInputStream, StandardCharsets.UTF_8)
        val stderrLog = IOUtils.toString(errorInputStream, StandardCharsets.UTF_8)
        outputInputStream.close()
        errorInputStream.close()
        logger.debug { ("command standard output: {}$stdoutLog") }
        logger.debug { ("command error output: {}$stderrLog") }
    }

    fun removeRestrictions(restrictedPdfPath: Path?, password: String?): Path? {
        logger.debug { "restricted filename = $restrictedPdfPath" }
        logger.debug { "password = $password" }
        val unrestrictedPdfPath: Path = restrictedPdfPath
            ?.resolveSibling(restrictedPdfPath.fileName.toString() + SUFFIX_PDF_UNRESTRICTED)!!
        logger.debug { "unrestricted filename = $unrestrictedPdfPath" }
        try {
            val processBuilder = buildProcess(restrictedPdfPath, password, unrestrictedPdfPath)

            // run command
            val process = processBuilder.start()
            readOutput(process)
        } catch (e: IOException) {
            logger.error(e) { "Something went wrong during converting PDF" }
            return null
        }
        return unrestrictedPdfPath
    }
}